lib/pka.c

```diff
@@ -1694,6 +1748,166 @@ int pka_modular_inverse(pka_handle_t   handle,
+                               uint16_t          curve_type,
+                               pka_operand_t*    point_x)
+{
+    pka_comparison_t rc;
+    uint32_t         idx;
+    uint8_t          big_endian, ls_byte, ms_byte;
+
+    big_endian = point_x->big_endian;
+    if (curve_type == PKA_CURVE_25519)
+    {
+        if (point_x->actual_len != 32)
+            return 1;
+
+        // We want to see if point_x (with a special adjustment of the
+        // most significant byte) is < the curve prime.
+        // First do a quick test
+        ls_byte = point_x->buf_ptr[big_endian ? 31 : 0];
+        if (ls_byte < 0xED) // 237 = 256 - 19
+            return 1;
+
+        // Loop over the bytes from least significant to most significant
+        // looking for a byte != 0xFF.  The most signifcant byte is special.
+        if (big_endian)
+        {
+            for (idx = 1; idx <= 30; idx++)
+                if (point_x->buf_ptr[31 - idx] != 0xFF)
+                    return 1;
+
+            ms_byte = point_x->buf_ptr[0];
+        }
+        else
+        {
+            for (idx = 1; idx <= 30; idx++)
+                if (point_x->buf_ptr[idx] != 0xFF)
+                    return 1;
+
+            ms_byte = point_x->buf_ptr[31];
+        }
+
+        ms_byte &= 0x7F;
+        return ms_byte != 0x7F;
+    }
+    else if (curve_type == PKA_CURVE_448)
+    {
+        if (point_x->actual_len != 56)
+            return 1;
+
+        // Quick test *TBD*
+        ms_byte = point_x->buf_ptr[big_endian ? 0 : 55];
+        if (ms_byte != 0xFF)
+            return 1;
+
+        rc = pka_internal_compare(point_x->buf_ptr, curve->p.buf_ptr, 56,
+                                  big_endian);
+        if (rc == PKA_LESS_THAN)
+            return 1;
+
+        return 0;
+    }
+    else
+        return 1;
+}
+
+int pka_mont_ecdh_canonicalize(pka_handle_t      handle,
+                               uint16_t          curve_type,
+                               ecc_mont_curve_t* curve,
+                               pka_operand_t*    point_x,
+                               pka_operand_t*    reduced_value)
+{
+    pka_result_code_t rc;
+    pka_operand_t     temp;
+    uint8_t           temp_buf[MAX_ECC_BUF];
+    int               is_canonical;
+
+    is_canonical = pka_is_mont_ecdh_canonical(curve, curve_type, point_x);
+    if (is_canonical)
+        return -1;
+
+    // Make a local copy of point_x first
+    memcpy(&temp, point_x, sizeof(temp));
+    temp.buf_ptr = &temp_buf[0];
+    memcpy(temp.buf_ptr, point_x->buf_ptr, point_x->actual_len);
+    if (curve_type == PKA_CURVE_25519)
+        temp.buf_ptr[31] &= 0x7F;
+
+    rc = pka_internal_subtract(&temp, &curve->p, reduced_value);
+    if (rc == RC_NO_ERROR)
+        return 0;
+    else
+        return -1;
+}
+
+int pka_mont_ecdh_mult(pka_handle_t      handle,
+                       void*             user_data,
+                       uint16_t          curve_type,

```

Comment:
Do you think we can encapsulate these curve parameters/information (i.e., curve type and whether the curve is canonical in ecc_mont_curve_t)?